### PR TITLE
Implement Rvalue::Len.

### PR DESCRIPTION
**Pull request by ptersilie**
Companion PR: https://github.com/softdevteam/ykrustc/pull/135

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -598,6 +600,21 @@ impl<TT> TraceCompiler<TT> {
         Location::Register(reg)
     }
 
+    fn c_len(&mut self, p: &Place) -> Location {
+        let dst = self.create_temporary();
+        let (loc, _) = self.place_to_location(p, true);
+        match loc {
+            Location::Register(src) | Location::Addr(src) => {
+                dynasm!(self.asm
+                    ; mov Rq(dst), [Rq(src) + 8]
```
Comment (vext01) [inline ykcompile/src/lib.rs:609]: This is access into a fat pointer, right? Should we comment this?

Comment (ptersilie) [reply ykcompile/src/lib.rs:609]: Yes, I was just thinking that too.

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -598,6 +600,21 @@ impl<TT> TraceCompiler<TT> {
         Location::Register(reg)
     }
 
+    fn c_len(&mut self, p: &Place) -> Location {
+        let dst = self.create_temporary();
+        let (loc, _) = self.place_to_location(p, true);
+        match loc {
+            Location::Register(src) | Location::Addr(src) => {
```
Comment (vext01) [inline ykcompile/src/lib.rs:607]: Can the `Location::Register` case arise for a fat pointer?

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:611]: Try this: 119bbb

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:609]: To be honest, I'm not sure. If `Len` is always preceded by a `Deref`, then no. But if we have a non-referenced array, and use `Len` on it, we might just get a `Register`. In that case, we could possibly even get a `Mem`, but we'll have to implement arrays first in order to test this.

Comment (vext01) [reply ykcompile/src/lib.rs:609]: So shall we remove `Location::Register` and let it crash? Then we can decide what to do later.

Comment (ptersilie) [reply ykcompile/src/lib.rs:609]: Let's do it.

Comment (ptersilie) [reply ykcompile/src/lib.rs]: Fixed in 70227c9

**Comment by ptersilie** [comment]
How about 008a9c0

**Comment by vext01** [comment]
Great. Please squash.

**Comment by ptersilie** [comment]
Squashed. Just waiting on bors to finish building ykrustc.

**Comment by vext01** [comment]
Ready?

**Comment by ptersilie** [comment]
Yup.

**Comment by vext01** [comment]
Formatting 😑

**Comment by ptersilie** [comment]
Okay to squash?

**Comment by vext01** [comment]
Go ahead.

**Pull request merged**